### PR TITLE
Fix for non US-ASCII characters in Content-Disposition header

### PR DESCRIPTION
**Pull request by PowerGamer1**
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | no |
| Breaks BC? | yes |
| Tests pass? | yes |
| Fixed issues | https://github.com/yiisoft/yii2/issues/12047, https://github.com/yiisoft/yii2/issues/10563, https://github.com/yiisoft/yii2/issues/11953 |

Fix for non US-ASCII characters in Content-Disposition header.

### CODE REVIEW

**Comment by PowerGamer1** [comment]
Fixing this without breaking BC would not be possible. The old code would put `basename($filePath)` into the HTTP header. Since `$filePath` can have arbitrary encoding (for ex., UTF-16 under Windows) it would produce the invalid HTTP header. Another problem of the old code - it would not allow to create the `Content-Disposition` header without specifying filename. Doing so would allow the browser to infer the filename from the URL. And the final problem fixed - now the UTF-8 `$attachmentName` would be properly encoded in the HTTP header.

**Comment by shirase** [comment]
Many people use ascii only. Your solution have problem for them and it breaks BC.

**Comment by PowerGamer1** [comment]
> Many people use ascii only. Your solution have problem for them 

No problems with ASCII at all.

> and it breaks BC.

If that is the price to fix broken HTTP headers then BC must be broken.

File: framework/web/Response.php [hunk 1]
```diff
@@ -547,23 +547,29 @@ public function sendStreamAsFile($handle, $attachmentName, $options = [])
...
         $disposition = $inline ? 'inline' : 'attachment';
         $headers->setDefault('Pragma', 'public')
             ->setDefault('Accept-Ranges', 'bytes')
             ->setDefault('Expires', '0')
-            ->setDefault('Cache-Control', 'must-revalidate, post-check=0, pre-check=0')
-            ->setDefault('Content-Disposition', "$disposition; filename=\"$attachmentName\"");
+            ->setDefault('Cache-Control', 'must-revalidate, post-check=0, pre-check=0');
+
+		if($attachmentName !== null) {
+			$headers->setDefault('Content-Disposition', $disposition . '; filename*=UTF-8\'\'' . rawurlencode($attachmentName));
```
Comment (samdark) [inline framework/web/Response.php]: Older browsers aren't recognizing `filename*` at all so `filename` should be always provided.

**Comment by samdark** [comment]
This solution won't work in older browsers.

**Pull request closed without merging**